File: apps/web/src/lib/schemas/investments.ts
import { z } from "zod";

export const investmentFormSchema = z.object({
  amount: z
    .string()
    .trim()
    .regex(/^\d+(\.\d{1,4})?$/, "Use a number with up to 4 decimal places.")
    .refine((v) => Number(v) > 0, "Must be positive."),
  notes: z.string().trim().max(500).optional(),
});
export type InvestmentFormValues = z.infer<typeof investmentFormSchema>;

export const investmentSchema = z.object({
  id: z.string().uuid(),
  amount: z.string(),
  currency: z.string(),
  status: z.string(),
  escrow_reference: z.string(),
  notes: z.string().nullable(),
  bank_account_id: z.string().uuid(),
  settled_at: z.string().nullable(),
  created_at: z.string(),
});
export type Investment = z.infer<typeof investmentSchema>;

export const investmentListSchema = z.object({
  items: z.array(investmentSchema),
  total: z.number().int(),
});
export type InvestmentList = z.infer<typeof investmentListSchema>;

File: apps/api/app/schemas/investments.py
"""Investment schemas."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class InvestmentCreateRequest(_Strict):
    """Body for ``POST /api/v1/investments``.

    Currency is implicitly the linked bank account's currency — we don't
    accept it from the client to avoid mismatches.
    """

    amount: Decimal = Field(gt=Decimal("0"), max_digits=20, decimal_places=4)
    notes: str | None = Field(default=None, max_length=500)


class InvestmentPublic(BaseModel):
    id: UUID
    amount: Decimal
    currency: str
    status: str
    escrow_reference: str
    notes: str | None = None
    bank_account_id: UUID
    settled_at: datetime | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InvestmentList(BaseModel):
    items: list[InvestmentPublic]
    total: int
